pkg/controllers/management/rkeworkerupgrader/upgrade.go

```diff
@@ -346,14 +335,112 @@ func (uh *upgradeHandler) upgradeCluster(cluster *v3.Cluster, nodeName string) e
 	return nil
 }
 
-func filterNodes(nodes []*v3.Node, expectedVersion int) ([]*v3.Node, map[string]*v3.Node, map[string]*v3.Node, map[string]*v3.Node, map[string]*v3.Node, int, int) {
-	var filtered []*v3.Node
-	notReady := map[string]*v3.Node{}
+func (uh *upgradeHandler) toUpgradeCluster(cluster *v3.Cluster) (bool, error) {
+	if v3.ClusterConditionUpgraded.IsUnknown(cluster) {
+		return true, nil
+	}
 
-	done, upgrading := 0, 0
-	toProcessMap, toPrepareMap, doneMap := map[string]*v3.Node{}, map[string]*v3.Node{}, map[string]*v3.Node{}
+	nodes, err := uh.nodeLister.List(cluster.Name, labels.Everything())
+	if err != nil {
+		return false, err
+	}
+
+	for _, node := range nodes {
+		if node.Status.NodeConfig == nil {
+			continue
+		}
+
+		if !workerOnly(node.Status.NodeConfig.Role) {
+			continue
+		}
+
+		if node.Spec.NodePlan == nil {
+			continue
+		}
+
+		nodePlan, err := uh.getNodePlan(node, cluster)
+		if err != nil {
+			return false, err
+		}
+
+		if planChangedForUpgrade(nodePlan, node.Spec.NodePlan.Plan) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (uh *upgradeHandler) prepareNode(node *v3.Node, toDrain bool, nodeDrainInput *v3.NodeDrainInput) error {
+	var nodeCopy *v3.Node
+	if toDrain {
+		if node.Spec.DesiredNodeUnschedulable == "drain" {
+			return nil
+		}
+		nodeCopy = node.DeepCopy()
+		nodeCopy.Spec.DesiredNodeUnschedulable = "drain"
+		nodeCopy.Spec.NodeDrainInput = &v3.NodeDrainInput{
+			DeleteLocalData:  nodeDrainInput.DeleteLocalData,
+			Force:            nodeDrainInput.Force,
+			Timeout:          nodeDrainInput.Timeout,
+			GracePeriod:      nodeDrainInput.GracePeriod,
+			IgnoreDaemonSets: nodeDrainInput.IgnoreDaemonSets,
+		}
+	} else {
+		if node.Spec.DesiredNodeUnschedulable == "true" || node.Spec.InternalNodeSpec.Unschedulable {
+			return nil
+		}
+		nodeCopy = node.DeepCopy()
+		nodeCopy.Spec.DesiredNodeUnschedulable = "true"
+	}
+
+	if _, err := uh.nodes.Update(nodeCopy); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (uh *upgradeHandler) processNode(node *v3.Node, cluster *v3.Cluster) error {
+	nodePlan, err := uh.getNodePlan(node, cluster)
+	if err != nil {
+		return fmt.Errorf("setNodePlan: error getting node plan for [%s]: %v", node.Name, err)
+	}
+
+	nodeCopy := node.DeepCopy()
+	nodeCopy.Spec.NodePlan.Plan = nodePlan
+	nodeCopy.Spec.NodePlan.Version = cluster.Status.NodeVersion
+	nodeCopy.Spec.NodePlan.AgentCheckInterval = agentCheckIntervalDuringUpgrade
```

Comment:
I'm modifying when to set this AgentCheckInterval, but that doesn't impact overall logic and PR ready for first  review